batch timeline fetch milestones using subgraph and github api

anynomousfriend opened the pull request:
## Description
[Screencast From 2025-11-04 16-59-33.webm](https://github.com/user-attachments/assets/a6e5845c-6de4-49ee-a2ab-ebc8e8006325)


Migrated Batch Timeline from direct on-chain hooks to subgraph-powered hooks for improved performance. Key changes:

- Integrated The Graph subgraph with GraphQL queries
- Enhanced timeline cards with stronger glassmorphism blur effect
- Added automatic caching and better type safety

This supersedes PR #42 which used direct blockchain queries. The new approach provides faster indexed queries, automatic caching, and correct event labeling.

## Additional Information

- [x] I have read the [contributing docs](/scaffold-eth/scaffold-eth-2/blob/main/CONTRIBUTING.md)
- [ ] This is not a duplicate of any existing pull request

## Related Issues

_Supersedes #42_

_Closes #33 
Your ENS/address: 0x729a2043F6702Ad76A62215A69BC00360E290268

anynomousfriend committed:
Feature: Add Batch Timeline component with kickoff, check-in, and graduation milestones Closes #33

anynomousfriend committed:
Refactor: Extract timeline data fetching into custom React hooks

- Created useCheckInEvents hook for on-chain check-in events
- Created useGraduationNFTEvents hook for graduation NFT mints
- Created useGitHubPREvents hook for merged pull requests
- Created useTimelineEvents combined hook that merges all three data sources
- Updated BatchTimeline component to accept events as props
- Updated page.tsx to use hooks for data fetching

anynomousfriend committed:
improve: enhance timeline loading UX

anynomousfriend committed:
enhance card blur styling for glass effect; add subgraph and update hooks to use subgraph

melanke commented on packages/nextjs/app/api/timeline/route.ts:
```diff
@@ -0,0 +1,168 @@
+      event: parseAbiItem("event CheckedIn(bool first, address indexed builder, address checkInContract)"),
+      fromBlock: 0n,
+      toBlock: "latest",
+    });
```
Using `getLogs` is fine for small use cases, but it doesn’t scale — it’s slow, limited in filters, and needs manual reorg handling. Over time, the RPC may stop responding to large or frequent queries, so using an indexer is safer and more reliable for production.

That's why Parzival is using an indexer on his PR: https://github.com/BuidlGuidl/batch21.buidlguidl.com/pull/44/files

anynomousfriend replied:
Hi @melanke, thanks for the feedback! 

I actually migrated away from using `getLogs()` in the API route. In this PR, I've moved to **The Graph subgraph** for querying check-in and graduation NFT events instead of direct RPC calls.

The current implementation uses:
- `useCheckInEventsFromSubgraph()` - queries The Graph subgraph via GraphQL
- `useGraduationNFTEventsFromSubgraph()` - also uses The Graph subgraph
- `useGitHubPREvents()` - queries GitHub API for merged PRs

The `app/api/timeline/route.ts` file is now **not being used** and can be deleted. All data fetching happens through the hooks using the subgraph.

Should I remove that unused API route file to clean up the codebase?

melanke replied:
I don't understand, the code is still using getLogs, what do you mean by "I actually migrated away from using gitLogs"?

Please, make another push so we can view your updated PR.

anynomousfriend replied:
You're right, my apologies. The `app/api/timeline/route.ts` was from an earlier implementation. When I refactored to subgraph-based hooks, this API route became redundant. I've now removed it in the latest commit.

Current implementation uses client-side hooks:
- `useCheckInEventsFromSubgraph()` 
- `useGraduationNFTEventsFromSubgraph()`
- `useGitHubPREvents()`

The API route is no longer needed. Just pushed the cleanup. Thanks for catching it!

Sahib-Aujla reviewed (changes requested):
<img width="1663" height="815" alt="image" src="https://github.com/user-attachments/assets/a58aa467-933e-45f7-ad06-6d8731a83fb6" />

Background mismatch for light mode

anynomousfriend committed:
fix: add blur and transparency effects in light mode; remove unused API route

melanke commented on packages/nextjs/components/BatchTimeline.tsx:
```diff
@@ -0,0 +1,271 @@
+};
+
+export const BatchTimeline = ({ events }: BatchTimelineProps) => {
+  const { theme } = useTheme();
```
You should use Tailwind's dark variant, you don't need this hook.

melanke commented on packages/nextjs/hooks/types.ts:
```diff
@@ -0,0 +1,28 @@
+      author?: string;
+      address?: never;
+      link?: string;
+    };
```
This is very repetitive, use a base type, something like this:
```suggestion
type BaseTimelineEvent = {
      date: string;
      title: string;
      description: string;
      link?: string;
    }
type VariantTimelineEvent =
  | {
      type: "on-chain-checkin";
      address?: string;
      author?: never;
    }
  | {
      type: "graduation-nft";
      address?: string;
      author?: never;
    }
  | {
      type: "pr-merged";
      author?: string;
      address?: never;
    };
export type TimelineEvent = VariantTimelineEvent & BaseTimelineEvent;
```

melanke commented on packages/nextjs/hooks/useGitHubPREvents.ts:
```diff
@@ -0,0 +1,79 @@
+    fetchGitHubPRs();
+  }, [owner, repo, token]);
+
+  return { events, isLoading, error };
```
You should use tanstack react-query, it's already in the project.

melanke commented on packages/nextjs/hooks/useTimelineEvents.ts:
```diff
@@ -0,0 +1,59 @@
+    checkInResult.error,
+    graduationNFTResult.error,
+    gitHubPRResult.error,
+  ]);
```
There's a very common pitfall when developing reactive applications: creating more states than you need.
    
You already have all the necessary variables in states or as constants, so you don't need a new state to store all the events. You simply need to transform what you have into a different format. Every time the other variables changes, the "all events" should change automatically, and no developer should be able to manually modify the "all events", which is currently possible in your code.
    
So instead of using `useState` and `useEffect`, use `useMemo` to transform the existing variables into "all events" in a new variable.
    
Try to always remember: if something can be built without `useState` and `useEffect`, it will be easier to maintain.
    
Take a look at the `useMemo` documentation and let me know if you have any questions.

anynomousfriend replied:
All the requested changes have been implemented. Please let me know if you need any further adjustments. Looking forward to your feedback!

melanke reviewed (changes requested):
Almost there

anynomousfriend committed:
code refactor: improve timeline data fetching and UI layout

melanke commented:
@anynomousfriend is it working on your machine? I'd just test it now for the first time and it keeps loading forever. Maybe there's a bug or is it that slow? How much does it take to load on your machine?

Also, please, add a link to the timeline page somewhere.

anynomousfriend committed:
fix: removed redundant sorting and optimized data merging in useTimelineEvents hook

anynomousfriend commented:
@melanke  I've done some tweaks to hopefully reduce loading time by removing redundant sorting operations and loading state shows only while subgraph queries are fetching in the timeline events fetching hook. Also, regarding the timeline page link—someone has already added it on the main repo in  `packages/nextjs/app/page.tsx` . The link to /timeline is fully implemented and merged.

Thanks!

melanke commented on packages/nextjs/components/BatchTimeline.tsx:
```diff
@@ -0,0 +1,229 @@
+                          bg-gradient-to-br ${style.cardBg}
+                          border-gray-200 dark:border-gray-700
+                          hover:border-gray-300 dark:hover:border-gray-600
+                          from-white/30 to-white/20 dark:from-gray-800/30 dark:to-gray-900/20
```
You should only have hover effects for interactive elements. This hover effect misleads the user into thinking they can click it to navigate.

melanke commented on packages/nextjs/components/BatchTimeline.tsx:
```diff
@@ -0,0 +1,229 @@
+"use client";
```
We don't need this restriction since this component doesn't use any client-only mechanism. Please remove this line so we can improve the app's performance by pre-rendering the component during the compile process.

melanke commented on packages/nextjs/app/timeline/page.tsx:
```diff
@@ -0,0 +1,136 @@
+"use client";
+
+import React, { useEffect, useState } from "react";
```
You don't need to explicitly import React in Next.js anymore, starting from Next.js 12.2 and React 17 and above. With these versions, Next.js automatically imports React when needed, so you can safely omit import React from "react" at the top of your files.

melanke commented on packages/nextjs/app/timeline/page.tsx:
```diff
@@ -0,0 +1,136 @@
+  "✨ Loading merged PRs from GitHub",
+];
+
+export default function TimelinePage() {
```
You might want to consider typing your page as NextPage since we are using TypeScript, as this ensures better integration and adds clarity for future maintainers by indicating that the component is a Next.js page.

melanke commented on packages/nextjs/components/BatchTimeline.tsx:
```diff
@@ -0,0 +1,229 @@
+                          whiteSpace: "normal",
+                          overflowWrap: "break-word",
+                          backdropFilter: "blur(12px)",
+                        }}
```
this can be added as tailwind classes

melanke commented on packages/nextjs/components/BatchTimeline.tsx:
```diff
@@ -0,0 +1,229 @@
+};
+
+export const BatchTimeline = ({ events }: BatchTimelineProps) => {
+  // const sorted = [...(events || [])].sort((a, b) => new Date(a.date).getTime() - new Date(b.date).getTime()); // not needed anymore as events are pre-sorted
```
It's bad practice to leave commented code. Other contributors might be confused about whether it should be uncommented or removed.

anynomousfriend replied:
Removed! Thanks for pointing this out—still learning 😁

melanke commented on packages/nextjs/hooks/useGitHubPREvents.ts:
This isn’t working. I’m only seeing the “On-Chain Check-in” and “Graduation NFT” events.
It’s preferable if you fix it, but if you can’t or don’t want to spend more time on this, we can simply remove the GitHub integration.

anynomousfriend replied:
I really want to make sure we get this working properly.

The GitHub PR integration is functioning correctly on my end, so I'd like to help figure out what might be different in your setup. 
[Screencast From 2025-11-07 12-35-18.webm](https://github.com/user-attachments/assets/ff288db4-0249-42d3-a718-11a77b73ca26)


Could I ask what variable name you used in your `.env.local` file for the GitHub token? The code specifically looks for `NEXT_PUBLIC_GITHUB_TOKEN`, so if you're using a different name (like `GITHUB_TOKEN`, `GH_TOKEN`, or another variation), it won't be able to find your API key from the `.env.local` file.

If you could check your `.env.local` and confirm whether it matches this format:

NEXT_PUBLIC_GITHUB_TOKEN=your_token_here

melanke replied:
Oh that's the problem, we can't set env variables for the Batch projects.
Since this is a `NEXT_PUBLIC` env variable (public for anyone to see), do you mind moving it to a hardcoded variable inside the code itself?
Thanks.

anynomousfriend replied:
I looked into it and found that both classic and fine-grained tokens get automatically revoked by GitHub within minutes if they're leaked in public code. I'm currently exploring an alternative: fetching merged PRs without any API token at all. Thanks.

anynomousfriend committed:
refactor: apply PR feedback - add NextPage type, remove React default import, optimize Tailwind classes

melanke commented on packages/nextjs/components/BatchTimeline.tsx:
```diff
@@ -0,0 +1,213 @@
+                          <h3
+                            className="text-lg md:text-xl font-semibold mb-2 break-words whitespace-normal text-gray-900 dark:text-white"
+                            title={event.title}
+                            style={{ wordBreak: "break-word", whiteSpace: "pre-wrap", overflowWrap: "break-word" }}
```
Oh, I miss this one, can you fix it as well?

anynomousfriend replied:
Done

anynomousfriend committed:
refactor: remove GitHub token requirement - use unauthenticated API requests

- Remove token parameter from useGitHubPREvents hook
- Remove Authorization header from GitHub API calls
- Update enabled condition to only check owner and repo
- Increase staleTime to 10 minutes for better rate limit compliance
- Works with 60 req/hour unauthenticated rate limit

anynomousfriend committed:
refactor: use tailwind classes for text wrapping instead of inline styles

melanke commented:
https://github.com/BuidlGuidl/batch21.buidlguidl.com/pull/45#issuecomment-3498301908
> Also, please, add a link to the timeline page somewhere.

https://github.com/BuidlGuidl/batch21.buidlguidl.com/pull/45#discussion_r2502988148
> this can be added as tailwind classes

anynomousfriend committed:
Link /timeline page to nav

anynomousfriend commented:
@melanke, Can you review all the changes? Thanks

> [#45 (comment)](https://github.com/BuidlGuidl/batch21.buidlguidl.com/pull/45#issuecomment-3498301908)
> 
> > Also, please, add a link to the timeline page somewhere. -- Done.

> [#45 (comment)](https://github.com/BuidlGuidl/batch21.buidlguidl.com/pull/45#discussion_r2502988148)
> 
> > this can be added as tailwind classes -- Done.
